src/lib/scroller.js

```diff
@@ -80,34 +88,61 @@ Scroller.prototype.requestRefresh = function(){
   if(this.dirty && this.console){
-    var top = this.console.scrollTop;
+    const top = this.console.scrollTop;
     if(this.sticky){
-      this.startPosition = Math.max(0, this.lines.length - this.minVisible);
+      this.startPosition = Math.max(this.lineOffset, this.lineCount() - this.visibleCount);
+      this.endPosition = this.lineCount();
     }
-    this.console.innerHTML = generateContent(this.lines, this.startPosition, this.lines.length, this.minVisible);
+    this.console.innerHTML = this._generateContent();
     if(this.jumpToBottom){
-      this.console.scrollTop = 2000;
+      this.console.scrollTop = 4000;
       this.jumpToBottom = false;
-    }else if(!this.sticky && this.startPosition > 0 && top === 0){
+    }else if(!this.sticky && this.startPosition > this.lineOffset && top === this.lineOffset){
       //cover the situation where the window was fully scrolled faster than expand could keep up and locked to the top
-      requestAnimationFrame(this.expand);
+      requestAnimationFrame(this.expandTop);
     }
     this.dirty = false;
   }
 };
 
-Scroller.prototype._expand = function(){
-  this.startPosition = Math.max(0, this.startPosition - this.minVisible);
-  this.sticky = false;
+Scroller.prototype._expandTop = function(){
+  this.startPosition = Math.max(this.lineOffset, this.startPosition - this.visibleCount);
   if(this.console){
-    var scrollHeight = this.console.scrollHeight;
-    var scrollTop = this.console.scrollTop;
+    this.sticky = false;
+    const scrollHeight = this.console.scrollHeight;
+    const scrollTop = this.console.scrollTop;
 
     // do an inline scroll to avoid potential scroll interleaving
-    this.console.innerHTML = generateContent(this.lines, this.startPosition, this.lines.length, this.minVisible);
-    var newScrollHeight = this.console.scrollHeight;
+    this.console.innerHTML = this._generateContent();
+    const newScrollHeight = this.console.scrollHeight;
     this.console.scrollTop = scrollTop + newScrollHeight - scrollHeight;
 
+    const oldEndPos = this.endPosition;
+    this.endPosition = Math.min(this.endPosition, this.startPosition + (this.visibleCount * 2));
+
+    this.dirty = oldEndPos !== this.endPosition;
+    if(this.dirty && !this.animateRequest){
+      this.animateRequest = requestAnimationFrame(this.refresh);
+    }
+  }
+};
+
+Scroller.prototype._expandBottom = function(){
+  this.endPosition = Math.min(this.lineCount(), this.endPosition + this.visibleCount);
+  if(this.console){
+    // add the new content to the bottom, then get scroll position to remove content
+    this.console.innerHTML = this._generateContent();
+    const scrollHeight = this.console.scrollHeight;
+    const scrollTop = this.console.scrollTop;
+
+    // update start position and render
+    this.startPosition = Math.max(this.lineOffset, Math.min(this.lineCount() - (this.visibleCount * 2), this.endPosition - (this.visibleCount * 2)));
+    this.console.innerHTML = this._generateContent();
```

Comment:
It's a two part update, where it adds new text at the bottom, then grabs the new scrollTop/scrollHeight (which forces the layout). After there it removes the lines above, but that causes the text position to jump. It then uses those scroll values to figure out the exact distance it needs to compensate.

I can't think of a way to do it without forcing two inline render calls, but it's pretty effective at avoiding scroll drift. The performance hit is pretty low too, since it's never rendering more than  50-100 lines at a time.